File: RightBicepCurlStrategy.py
from ExerciseAnalysisStrategy import IExerciseAnalysisStrategy
from AngleCalculator import AngleCalculator
import mediapipe as mp

class RightBicepCurlStrategy(IExerciseAnalysisStrategy):

    mp_pose = mp.solutions.pose.PoseLandmark

    def __init__(self, landmarks):
        
        self.hip_r = landmarks[self.mp_pose.RIGHT_HIP.value]

        self.shoulder_r = landmarks[self.mp_pose.RIGHT_SHOULDER.value]

        self.wrist_r = landmarks[self.mp_pose.RIGHT_WRIST.value]
        
        self.elbow_r = landmarks[self.mp_pose.RIGHT_ELBOW.value]

        self.r_elbow_angle = AngleCalculator.calculate_landmarks_angle(landmarks,
                                                                            self.mp_pose.RIGHT_SHOULDER,
                                                                            self.mp_pose.RIGHT_ELBOW,
                                                                            self.mp_pose.RIGHT_WRIST)
        
        self.r_shoulder_angle = AngleCalculator.calculate_landmarks_angle(landmarks,
                                                                            self.mp_pose.RIGHT_ELBOW,
                                                                            self.mp_pose.RIGHT_SHOULDER,
                                                                            self.mp_pose.RIGHT_HIP)
        
        self.r_shoulder_wrist = AngleCalculator.calculate_horizontal_distance(self.shoulder_r, self.wrist_r, 3)

    def correct_form(self):
        bicep_guide = ""

        if 20 < self.r_elbow_angle < 60 and 10 < self.r_shoulder_angle < 20 and 0.1 >= self.r_shoulder_wrist > -0.1 :
            bicep_guide = 'Keep Going!'

        else:
            #Conditon 1
            if 20 > self.r_elbow_angle:
                bicep_guide += 'Lower your wrist\n'
            if 60 < self.r_elbow_angle:
                bicep_guide += 'Raise your wrist\n'    
            #Condition 2
            if 0.1 < self.r_shoulder_wrist:
                bicep_guide += 'Move your wrist to the Left\n'
            if -0.1 > self.r_shoulder_wrist:
                bicep_guide += 'Move your wrist to the Right\n'
            # #Condition 3
            #if 10 > self.r_shoulder_angle:
             #   bicep_guide += 'Raise your elbow\n'
            #if 20 < self.r_shoulder_angle:
              #  bicep_guide += 'Lower your elbow\n'    

        print(self.r_shoulder_wrist)
        return bicep_guide

File: js/strategies/rightBicepCurlStrategy.js
import { IExerciseAnalysisStrategy } from './exerciseAnalysisStrategy.js';
import { AngleCalculator } from '../angleCalculator.js';
import { POSE } from '../landmarks.js';

export class RightBicepCurlStrategy extends IExerciseAnalysisStrategy {
    constructor(landmarks) {
        super(landmarks);

        this.hip_r = landmarks[POSE.RIGHT_HIP];
        this.shoulder_r = landmarks[POSE.RIGHT_SHOULDER];
        this.wrist_r = landmarks[POSE.RIGHT_WRIST];
        this.elbow_r = landmarks[POSE.RIGHT_ELBOW];

        this.r_elbow_angle = AngleCalculator.calculateLandmarksAngle(landmarks,
            POSE.RIGHT_SHOULDER, POSE.RIGHT_ELBOW, POSE.RIGHT_WRIST);

        this.r_shoulder_angle = AngleCalculator.calculateLandmarksAngle(landmarks,
            POSE.RIGHT_ELBOW, POSE.RIGHT_SHOULDER, POSE.RIGHT_HIP);

        this.r_shoulder_wrist = AngleCalculator.calculateHorizontalDistance(this.shoulder_r, this.wrist_r, 3);
    }

    correctForm() {
        let bicepGuide = '';

        // The shoulder-angle gate is disabled here to match LeftBicepCurlStrategy;
        // with it enabled the right arm was far harder to satisfy than the left.
        if (20 < this.r_elbow_angle && this.r_elbow_angle < 60 &&
            0.1 >= this.r_shoulder_wrist && this.r_shoulder_wrist > -0.1) {
            bicepGuide = 'Keep Going!';
        } else {
            // Condition 1
            if (20 > this.r_elbow_angle) {
                bicepGuide += 'Lower your wrist\n';
            }
            if (60 < this.r_elbow_angle) {
                bicepGuide += 'Raise your wrist\n';
            }
            // Condition 2
            if (0.1 < this.r_shoulder_wrist) {
                bicepGuide += 'Move your wrist to the Left\n';
            }
            if (-0.1 > this.r_shoulder_wrist) {
                bicepGuide += 'Move your wrist to the Right\n';
            }
        }

        return bicepGuide;
    }
}
